[이수연] TodoList with Ajax

2SOOY opened the pull request:
## 📌 구현 사항

- [x] todoItem 불러오기
- [x] todoItem 추가하기
- [x] todoItem 삭제하기
- [x] todoItem complete하기
- [x] todoItem contents 내용 수정하기
- [x] user별 투두리스트 불러오기
- [x] 데이터를 불러오기전 로딩바를 이용해, 사용자가 데이터가 불러와지고 있다는 것을 보여줍니다.
- [x] github issue에서 라벨을 붙이는 것처럼, 우선순위에 따라서 badge를 추가합니다.

## 📌 구조

```sh
├── index.html
├── css
│   └── app.css
└── js
    ├── index.js
    ├── Components
    │   ├── App.js             # state 관리 및 모든 하위 컴포넌트 관리
    │   ├── LoadingView.js     # api 요청시 사용자에게 보여줄 로딩 컴포넌트
    │   ├── TodoClearButton.js # 모든 todo item을 삭제하는 컴포넌트
    │   ├── TodoCount.js       # 활성화 된 tab에 따라 todo item 개수를 보여주는 컴포넌트
    │   ├── TodoInput.js       # 사용자의 입력을 받아 todo item을 추가하는 컴포넌트
    │   ├── TodoList.js        # 선택된 사용자의 todo list를 보여주고, todo item 조작을 위한 컴포넌트
    │   ├── TodoTab.js         # 전체, 미완료, 완료에 대한 tab을 제공하는 컴포넌트
    │   ├── TodoTitle.js       # 선택된 유저의 이름을 보여주는 title 컴포넌트
    │   └── UserList.js        # 서버 내 저장된 유저들의 목록 제공 및 유저 선택을 위한 컴포넌트
    └── utils
        ├── api.js
        ├── constant.js
        ├── template.js        # 컴포넌트에서 rendering될 html markup을 모은 유틸
        └── validator.js
```

## 📌 배운 것

- 모든 API 요청 시 중복되는 fetch API 구조에 대해 api라는 객체로 관리하기
    + 호준님 api 처리 참고 👍 #1 
- `<selct>` 태그 및 `change 이벤트`
- Todo Item 변경 시 `.select()` 메소드
- 각 컴포넌트에서 사용하는 `html markup`에 해당하는 부분 `template`으로 모듈화
    + 동준님 PR 2. template 참고 👍 https://github.com/next-step/js-todo-list-step2/pull/2#pullrequestreview-438166309
## 📌 고민한 것

- App의 state를 어떻게 설계할 것인가 ?
  ```js
  this.state = {
    user: {
      name: '',
      todos: [],
    },
    users: [],
    selectedTab: '',
  };
  ```
- 각각의 하위 컴포넌트는 어떤 state를 가져야 하는가 ?
- api 요청은 어디에서 진행해야 하는가 ?
   + 각 컴포넌트 이벤트 발생 시 => App 컴포넌트로 전달받은 메소드를 통해 api 요청
- 로딩 시점은 어느 곳이 적절한지 ?
   + 유저 목록과 선택된 유저의 todoIList를 불러오는 api 처리시 로딩 진행
- 각종 발생할 수 있는 예외사항
  - 선택된 유저가 없는 상태에서 어플리케이션 조작

## 📌 생각해볼 것
`어플리케이션 로직`
+ 현재까지 설계한 로직
    - 최초 어플리케이션 실행 -> 유저 목록 불러옴 -> 사용자가 유저 선택 -> todoList 조작
+ `유저를 등록`할 수 있는 로직을 추가해서 새로운 유저도 어플리케이션을 사용할 수 있도록 하기

`어플리케이션 설계 구조`
+ 이번 주차 미션을 진행하면서 내가 설계할 수 있는 어플리케이션의 구조가 획일화 되어있다는 생각이 들었음
+ 지난주 미션에서 사용했던 코드 대부분을 거의 복붙함
+ 다른 분들의 PR을 적극적으로 보면서 어플리케이션의 더 나은 구조에 대해서 생각해보기

`코드의 중복 최소화`
+ App 컴포넌트의 대부분의 api요청을 처리하는 메소드의 구조가 동일함
+ 어떻게 해당 메소드를 공통화 시킬 수 있을지 고민해보기

2SOOY committed:
Init - TodoList with Ajax 🎉

- 컴포넌트의 의미를 명확히 하기 위해  index.html 몇몇 className 수정
- README.md 추가
- index.js 초기화

2SOOY committed:
Add - template.js : split html markup

- html내의 컴포넌트에서 사용될 item들을 template으로 분리함
- title 분리
- userList의 userButton 분리
- todoItem 분리
- todoItem의 priority 분리
- loadingView 분리
- todoCount 분리

2SOOY committed:
Add - api.js : todoList API

- 유저 목록 조회 : fetchUserList
- 유저 todoList 조회 : fetchUserTodo
- 유저 todo 추가 : addTodo
- 유저 todo 토글, 삭제, 내용 변경 : toggleTodo, deleteTodo, editTodoContent
- 유저의 모든 todo 목록 지우기 : deleteAllTodo
- BUG : 유저 todo의 우선순위 변경 500 ERROR 발생 (원인 잘 모르겠음)
- option 객체를 이용하여 각 요청 메소드에서 필요한 객체 정보를 따로 관리
- request 함수를 통해 fetch 및 response를 받는 부분 공통화
- api 객체 내에서 각종 api 함수를 제공하도록 설계함

2SOOY committed:
Add - TodoTitle component

- Todo App의 title에 대한 컴포넌트 설계
- state로 name을 갖음
- todoTitleHTML 템플릿 적용

2SOOY committed:
Add - constant.js : DOM selector

- DOM 객체에 대한 css 선택자 상수로 관리

2SOOY committed:
Add - TodoTitle to App component

- App 컴포넌트 내에 TodoTitle 컴포넌트 적용

2SOOY committed:
Add - constant.js : node name constant

- userList 컴포넌트의 클릭 이벤트 'BUTTON' 상수화

2SOOY committed:
Add - UserList component

- 존재하는 유저의 목록을 보여주는 컴포넌트
- 유저를 클릭하여 해당 유저의 todoList를 볼 수 있도록 해줌
- state로 users(유저목록)과 name(선택된 유저 이름)을 갖음
- userButton에 대한 template 코드 변경
- 선택된 유저의 버튼을 활성화 시키도록 적용

2SOOY committed:
Add - UserList to App component

- onChangeUser 메소드 추가
- UserList 컴포넌트 click 이벤트 발생 시, users와 user.name, user.todos 갱신

2SOOY committed:
Add - TodoList component : render

- 선택된 유저의 todoList를 보여주는 컴포넌트
- state로 유저이름, 해당 유저의 투두리스트, 현재 선택된 탭을 갖음
- todoList에 따라 렌더링 하는 로직 구성
- template.js에서 todoItem에 대한 템플릿 변경
- todoItem의 우선순위에 따라 선택팝업 or 우선순위 label 부여
- 우선순위 값에 대한 상수로 관리 PRIORITY

2SOOY committed:
Add - TodoList component : edit, toggle, remove

- click 이벤트에 대해 토글 or 삭제 처리
- 토글하는 과정에서 toggleTodo API 호출
- 삭제하는 과정에서 deleteTodo API 호출
- dblclick 이벤트에 대해 수정 모드
- 내용 수정 후 enter를 입력하면 editTodo API 호출
- 내용 수정 후 ESC 누를 시 기존 content 유지
- state에 isEditing 추가함 => 동시에 여러가지의 todo변경을 막기 위함
- template.js에서 completed 되었을 때 style 추가
- constant.js에서 KEY, NODE, MESSAGE 등 여러가지 상수 추가

2SOOY committed:
Add - edit, delete, toggle todo logic in App component

- TodoList에서 발생한 이벤트에 대해서 App컴포넌트의 메소드로 처리하도록 함
- 해당 메소드 내에서 API 호출이 이뤄짐
- 이후 API 호출 결과를 바탕으로 state를 변경함
- edit, delete, toggle 전부 변화한 state를 받아오는 logic이 중복됨
- 따라서, fetchState 라는 메소드 추가 및 각 메소드에 적용

2SOOY committed:
Add - TodoInput component : add todo

- 선택한 사용자의 todo를 추가하기 위한 컴포넌트
- state로 userName이 필요함
- api 요청시 user name이 필요하기 때문에
- 해당 컴포넌트 역시 name이라는 state 필요
- name이 없어도 설계가 가능하긴 하지만, App컴포넌트의 각종 메소드의 통일을 위해 state 부여
- 선택된 사용자가 없이 todo를 추가할 경우 처리 - NO_SELECTED_USER

2SOOY committed:
Add - TodoInput component in App

- 새로운 todo를 추가할 있는 method onAddTodo 추가
- todoInput에게 user name을 state로 넘겨줌

2SOOY committed:
Add - TodoCount component : display todo count

- 사용자의 todoList의 개수를 보여주는 컴포넌트
- 선택된 탭에 따라 todoList의 개수를 렌더링함

2SOOY committed:
Add - TodoCount component in App

- todos와 slectedTab을 state로 넘겨줌
- selectedTab에 따라 todos를 필터링하여 todoCount를 보여줌

2SOOY committed:
Add - TodoTab component : select all, active, completed todo

- 모든 todo, 해야할 todo, 완료된 todo를 선택할 수 있는 컴포넌트
- state로 selectedTab만 갖음
- 유일하게 api 요청을 사용하지 않음
- 서버 상에 selectedTab에 대한 정보는 없고, App자체의 state이기 때문에
- tab클릭 시 App의 onChangeTab 메소드 실행

2SOOY committed:
Add - TodoTab component in App

- index.html내 tab의 className 추가 all, active, completed
- TodoTab 컴포넌트를 위한 onChangeTab 메소드 추가
- TodoTab에서 클릭 이벤트 발생 시 App state의 selectedTab 변경

2SOOY committed:
Add - rendering selected tab in TodoList component

- getSelectedTodos라는 메소드를 통해 현재 선택된 TAB의 todos 정보만 갖고 오도록 함
- 이후 선택된 탭에 따라 가져온 정보를 render 시킴

2SOOY committed:
Fix - change todo priority api

- api 요청 주소를 잘못 입력했음
- 마지막에 /priority를 입력하지 않았서 500 ERROR 발생

2SOOY committed:
Add - change todo priority in TodoList component

- todoList 컴포넌트의 change 이벤트 발생 처리 => 우선순위 변경
- App의 onChangePriority 메소드를 통해 priority 변경 api 호출
- change 이벤트 발생 시 value에 따라 렌더링될 todoItem이 결정됨

2SOOY committed:
Add - TodoClearButton component : delete all todo items

- 모든 todo item을 삭제하는 컴포넌트
- App 컴포넌트에서 onDeleteAllTodo 메소드를 통해 모든 item을 삭제하는 API 요청

2SOOY committed:
Add - <ul> for LoadingView component

- <ul>태그의 style 변경
- margin, padding, line style 없애기

2SOOY committed:
Add - LoadingView component for api request & response

- api요청 시 사용자에게 보여줄 loading 컴포넌트
- 선택된 유저의 todoList와 유저 목록에 대한 api 요청 발생 시 로딩 처리
- state로 isLoading을 갖음

2SOOY committed:
Add - LoadingView component in App

- fetchState 동작 시 사용자게 로딩창 보여주도록 설계
- loading에 대한 state는 App 컴포넌트에서 따로 관리하지 않음
- 왜냐하면 해당 state는 LoadingView만 사용하는 state이기 때문

2SOOY committed:
Doc - update README.md

- 어플리케이션 전체적인 구조 사진 첨부
- 스터디 진행하면서 고민했던 부분 정리

2SOOY committed:
Update - TodoClearButton clicked logic : no selected user

- 어플리케이션 시작시 선택된 유저가 없을 때, 버튼 클릭 시 안내 메시지 로직 추가
- 불필요한 api요청을 막음

2SOOY committed:
Add - validator : check app state, target

- 올바른 HTML 요소인지 체크하는 함수 checkTarget
- app state의 적절한 프로퍼티를 전부 갖고 있는지 체크하는 함수 checkAppState

2SOOY committed:
Add - validation logic in components

- 생성된 컴포넌트에 대해서 target 검사 진행
- App의 state 변경 시, 프로퍼티 검사 진행

2SOOY committed:
Refactor - remove not used export keyword

- api.js, template.js 내의 export가 필요없는 것들 키워드 제거
- UserList의 reduce 부분 정리

wooooooood commented:
안녕하세요~~ 이번 미션 리뷰팀원 나무입니당
js초보라 수연님 코드 보면서 저를 더 많이 되돌아볼수 있는 좋은 기회였던 것 같아요 ㅎㅎㅎ 설명해주시는 구조를 엄청 깔끔하게 작성해주셔서 코드 읽기 전에 목차를 보는 느낌,,! 👍 수고하셨습니당!!!👏👏👏

wooooooood commented on js/Components/App.js:
```diff
@@ -0,0 +1,203 @@
+      selectedTab: CLASS_NAME.ALL,
+    };
+
+    this.state.users = await api.fetchUserList();
```
제가 아직 `async` `await`에 익숙하지 않아서 이부분에 많은 도움을 얻고 갑니다..ㅎㅎ

2SOOY replied:
도움이 되셨다니 감사합니다 :)

wooooooood commented on js/Components/App.js:
```diff
@@ -0,0 +1,203 @@
+    this.todoClearButton.setState(this.state.user.name);
+  };
+
+  this.init();
```
핫 제가 init을 사용해본 적이 없는데 이렇게 상태관리하는게 유용해보이네요 👀👍

2SOOY replied:
저는 함수의 body 내에서 이것저것 관리하는 것 보다, 
`init`이라는 메소드를 정의하여 관리하는 부분만 묶어 놓는게 보기 편해서 관습적으로 사용하고 있습니다 😄

SkynI25 replied:
전체적으로 코드가 보기 깔끔하게 잘 작성하신 것 같습니다 :) 많은 자극이 되네요

wooooooood commented on js/Components/TodoList.js:
```diff
@@ -0,0 +1,155 @@
+    this.render();
+  };
+
+  this.bindEvents = () => {};
```
이 부분은 40번 줄과 어떤 차이가 있는지 여쭤봐도 될까욤??

2SOOY replied:
제거해야 할 부분이군요,,
꼼꼼히 봐주셔서 감사합니다 👍

wooooooood commented on js/utils/api.js:
```diff
@@ -0,0 +1,95 @@
+  }),
+};
+
+const isSuccessResponse = (status) => !!(status >= 200 && status < 300);
```
오 이런 방법 너무 좋은것같아요!! 👏👏👏

2SOOY replied:
감사합니다 :)
추가로 `fetch API`를 통해 받은 응답이 response 라면,
`response.ok`를 통해 간단하게 확인하는 방법도 있습니다 :)


```js
let response = await fetch(url);

if (response.ok) { // HTTP 상태 코드가 200~299일 경우
  // 응답 몬문을 받습니다.
  let json = await response.json();
} else {
  alert("HTTP-Error: " + response.status);
}
```
https://ko.javascript.info/fetch

wooooooood commented on js/utils/constant.js:
```diff
@@ -0,0 +1,63 @@
+  SELECTED_TAB: 'selectedTab',
+};
+
+export const MESSAGE = {
```
이부분은 그저.. 감탄했습니다..

2SOOY committed:
Style - remove duplicated method : bindEvent

SkynI25 commented on js/Components/TodoClearButton.js:
```diff
@@ -0,0 +1,34 @@
+import { NODE, MESSAGE } from '../utils/constant.js';
+import { checkTarget } from '../utils/validator.js';
+
+function TodoClearButton({ $target, userName, onDeleteAllTodo }) {
```
안녕하세요 수연님? :) 혹시 모든 아이템을 삭제하는 컴포넌트를 따로 정의하신 이유가 있으신가요? :)

2SOOY replied:
우선 html 구조를 살펴보면, 다음과 같습니다 :)
```html
 <div class="count-container">
    <span class="todo-count"></span>
    <ul class="todo-tab"></ul>
    <button class="clear-all"></button>
 </div>
```
앞서 `todo-count`와 `todo-tab`이 각각의 역할에 따라 컴포넌트를 분리했기 때문에,
이와 같은 규칙에 따라 `clear-all`도 `할일을 모두 삭제한다`는 역할만 수행하는 컴포넌트로 설계하였습니다 :)

SkynI25 commented on js/Components/TodoInput.js:
```diff
@@ -0,0 +1,45 @@
+    if (!this.name) {
+      alert(MESSAGE.NO_SELECTED_USER);
+      e.target.value = '';
+      return;
```
이름이 없을 경우에 대한 에러 처리를 하셨네요. 보면서 배웁니다 :) 👍

SkynI25 commented on js/Components/TodoInput.js:
```diff
@@ -0,0 +1,45 @@
+
+    const keyword = e.target.value.trim();
+
+    if (!keyword.length) {
```
내용을 입력하지 않았을 경우에 대한 처리도 하셨네요! 중요한 처리인거 같습니다 :)

SkynI25 commented on js/Components/TodoList.js:
```diff
@@ -0,0 +1,153 @@
+
+    const todoItem = e.target.closest('li');
+    todoItem.classList.add(CLASS_NAME.EDITING);
+    todoItem.querySelector(SELECTOR.EDIT).select();
```
select() 메서드를 사용하셨네요! 이러면 커서를 뒤로 하는 작업도 필요없겠네요. 👍 
https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/select
그리고 quertySelector를 이용해서 input 태그를 지정한 방법 또한 배우고 갑니다 :)

2SOOY replied:
감사합니다 !
이전 기수 분들의 PR을 참고하면서 `커서를 뒤로 옮기는 방법`에 대해 고민했는데,
마침 `select`라는 좋은 메서드가 있어서 활용하게 됐습니다 :)

SkynI25 commented on js/Components/TodoList.js:
```diff
@@ -0,0 +1,153 @@
+    switch (key) {
+      case KEY.ENTER:
+        if (!editContent.length) {
+          alert(MESSAGE.NO_INPUT_KEYWORD);
```
혹시 어떤 의도로 alert를 사용하셨는지 여쭤봐도 될까요? 개인적으론 alert가 뜨는게 사용자 입장에선 불편할수도 있겠다 생각했거든요, 이에 대한 수연님의 의견이 궁금합니다 :)

2SOOY replied:
앞서 Todo Input에 `할일을 입력하지 않았을 경우`와 같은 맥락입니다 !
투두리스트의 item 내용을 변경할 때, `입력된 내용이 없을 경우` (item의 컨텐츠가 비었음)에 대한 처리를 진행하였습니다 :)

`alert`을 활용하는 방법은 사용자 입장에서 불편한 경험이라는 것은 동의합니다 :)
하지만, alert이 `가장 간단하게 에러를 보여줄 수 있는 방법`이여서 사용했습니다 ㅜ

ERROR 표시를 위한 별도의 Error UI 컴포넌트를 만드는 방법도 고려해보겠습니다 👍 
@SkynI25 님 덕분에 사용자의 경험에 대해서도 고려해 보는 생각을 가질 수 있었습니다. 세심한 리뷰 감사합니다 !! 😁

SkynI25 commented on js/Components/UserList.js:
```diff
@@ -0,0 +1,51 @@
+  };
+
+  this.createUserListHTML = (users, selectedName) => {
+    return users.reduce((html, user) => {
```
와우..template과 reduce를 사용해서 마크업 텍스트를 완성하셨네요 👍  인상적입니다!! :)

2SOOY replied:
저도 template을 사용하는 부분은 다른 분들의 PR을 참고해서,
많은 도움이 됐습니다 👍👍👍

ganeodolu commented on js/Components/App.js:
```diff
@@ -0,0 +1,203 @@
+  };
+
+  // UserList
+  this.onChangeUser = async (userName) => {
+    const newState = await this.fetchState(userName);
+
+    this.setState(newState);
+  };
```
`const newState = await this.fetchState(userName);` 이 부분이 각 메소드마다 공통으로 사용되고 있는 것 같아서
this.setState에 이 부분을 넣으면 중복을 줄일 수 있지 않을까? 해서 생각해봤습니다.
위 부분을 아래와 같이 변경하고,
```javascript
this.onChangeUser = async (userName) => {
    this.setState(userName)
  };
```
`this.setState`를 다음과 같이 수정하면
```javascript
  this.setState = async (userName) => {
    this.state = await this.fetchState(userName);
    checkAppState(this.state);
     ....
```
각 메소드별로 `const newState = await this.fetchState(userName);`을 줄일 수 있을 것 같아요 :frog:

2SOOY replied:
@ganeodolu 님의 좋은 의견 감사합니다 :) 
정말 고민을 많이 했던 부분 중에 하나인데요,

저는 `setState` 메소드에 대해 다음과 같은 정의를 내려 사용했습니다.
> 오직 `변경될 state 객체`만 받아, 내부에서는 별도의 API처리 없이 `해당 컴포넌트의 state를 update`하고 그에 맞게 `render` 하는 메소드이다.

따라서 해당 규칙을 지키기 위해, 각 API처리에 대한 메소드에서 중복이 발생할 수 밖에 없었던 것 같아요 😢

호준님 덕분에 조금 더 유연하게 사고하는 방법에 대해 배웠습니다 👍👍

ganeodolu replied:
규칙에 맡게 통일성을 유지하려고 하셨는데, 제가 단순히 중복만 줄이려고 생각했었네요. :sweat_smile: 저도 나중에 `setState`를 만들때 다시 생각해보게 될 것 같습니다. 답변 감사합니다.

ganeodolu commented on js/index.js:
```diff
@@ -0,0 +1,10 @@
+import { SELECTOR } from './utils/constant.js';
+
+try {
+  document.addEventListener('DOMContentLoaded', () => {
```
이 부분을 사용하면 어떤 장점이 있는지 궁금합니다.
정상적으로 DOM이 로드되지 않을 경우를 방지하기 위해 사용하는 것일까요 :question:

eastjun-dev replied:
자바스크립트 파일을 body가 끝나기전에 선언해주신다면 굳이 없어도 되는 부분일 것 같아 보여요 😄 
혹시 어떤 의도로 작성하신건지 여쭤봐도 괜찮을까요? 🤔

2SOOY replied:
사실 이 부분은 예전에 js와 관련된 강의를 들으면서,
"`DOMContentLoaded`이벤트가 발생했을 때 로직을 처리하는 방법도 있구나" 라는 것을 알게되었습니다.

그래서 별다른 의도없이 한 번 사용해보고자 적용했습니다 😢

두 분 덕분에 `DOMContentLoaded`이벤트에 대해서 자세히 알아봐야겠다는 생각이 들었습니다! 
세심한 리뷰 감사합니다 👍

ganeodolu commented on js/Components/TodoTab.js:
```diff
@@ -0,0 +1,51 @@
+
+    const clickedTab = e.target.className;
+    if (clickedTab.includes(CLASS_NAME.SELECTED)) return;
+
```
36번째 줄 `this.render` 에서 for 안의 조건문을 줄이기 위해 `this.render`내용을 아래와 같이 변경하여
```javascript
for (const tab of this.$tabs) {
    tab.classList.remove(CLASS_NAME.SELECTED);
}
e.target.classList.add(CLASS_NAME.SELECTED);
```
`this.onClick` 안에 넣는 방법도 괜찮지 않을까 생각합니다. :whale:

2SOOY replied:
click 이벤트 발생 시 `e.target`을 이용할 수 있으니깐, 
classList를 변경하기 훨씬 수월해지는 군요 ! 덕분에 render에 대한 메소드도 정의할 필요가 없어지네요 :)
이 방식 정말 괜찮은 것 같아요 !! 호준님 감사합니다 👍👍

ganeodolu commented on js/utils/validator.js:
```diff
@@ -0,0 +1,41 @@
+import { MESSAGE, APP_STATE } from './constant.js';
+
+export const checkTarget = ($target) => {
+  if (!$target instanceof HTMLElement) {
```
`if (!($target instanceof HTMLElement))` 괄호 바깥에 `!`를 사용하는 것이 더 정확하지 않을까 생각합니다.

2SOOY replied:
그렇군요 ! 이름만 `checkTarget`으로 짓고 실질적으로 제대로된 역할을 하지 못하고 있었네요ㅜ
섬세한 리뷰 감사합니다 👍

ganeodolu commented on js/Components/UserList.js:
```diff
@@ -0,0 +1,51 @@
+  this.onClick = (e) => {
+    if (e.target.nodeName !== NODE.BUTTON) return;
+
+    const selectedName = e.target.innerText;
```
꼭 필요한 경우가 아닌 경우에는 `innerText` 보다 `textContent`를 사용하시는 것을 추천드립니다
https://developer.mozilla.org/ko/docs/Web/API/Node/textContent#innerText%EC%99%80%EC%9D%98_%EC%B0%A8%EC%9D%B4%EC%A0%90

2SOOY replied:
사실 `innerText와 textContent의 차이`에 대해서 잘 모르고 있었는데,
자료까지 첨부해주시다니 정말 감사합니다 👍
덕분에 앞으로는 목적에 맞게 잘 사용할 수 있을 것 같아요 !

현재는 markup내의 style요소를 제외한 `text성분만 필요`해서 innerText를 활용하는 점이 더 편하게 사용할 수 있을 것 같아요 !
textContent를 사용하면 style요소를 없애는 별도의 후처리 작업이 필요할 것 같아요 :)

ganeodolu commented:
@2SOOY 님 코드, 커밋메세지와 다른 분 PR에 남긴 리뷰들을 보면서 배울 점이 많다고 느꼈습니다. 많은 고민하시면서 구조를 짜신 것 같아요. 그리고 closed된 PR까지 찾아서 api를 활용해주시다니 뿌듯하네요. 리뷰 몇가지 남겨드렸는데 도움이 되셨으면 좋겠습니다. 미션3도 기대하겠습니다. :frog:

eastjun-dev commented on .prettierrc:
```diff
@@ -0,0 +1,7 @@
+{
```
`.prettierrc` 사용까지 👍

eastjun-dev commented on README.md:
```diff
@@ -0,0 +1,84 @@
+# ☕️ 코드리뷰 모임 - Black Coffee
```
우와.... README에 요구사항 분석하신 부분 너무 멋지신데요!?!? 👍 💯

2SOOY replied:
정말 감사합니다 😄

eastjun-dev commented on css/app.css:
```diff
@@ -55,32 +61,32 @@ body {
 }
 
-.todoapp {
+.todo-body {
```
필요에 맞게끔 class 이름 수정해주신 것도 너무 좋은 것 같아요 😄

eastjun-dev commented on js/Components/App.js:
```diff
@@ -0,0 +1,203 @@
+    checkTarget($target);
+    this.$target = $target;
+
+    this.state = {
```
state관리하는 필드를 하나에서 관리하셨군요! 나중에 이 state가 커진다면 해당 부분만 따로 객체로 분리해도 될 것 같아 인상적인 접근인것 같아요 👍

2SOOY replied:
`확장성`을 고려했을 때, state를 어떻게 분리해야 하는지 리뷰주신 것에 정말 감사합니다 👍

eastjun-dev commented on js/Components/App.js:
```diff
@@ -0,0 +1,203 @@
+      selectedTab: CLASS_NAME.ALL,
+    };
+
+    this.state.users = await api.fetchUserList();
```
비동기 통신의 경우 결과가 언제나 success일 수 없기 때문에 try~catch로 에러가 난 상황도 처리해주시면 더 안정적인 코드가 될 것 같아요 😄

2SOOY replied:
실질적으로 API를 호출하는 부분에서 전부 `try...catch`가 빠져있군요ㅜ
비동기 통신에 대한 `try...catch`처리 적용하겠습니다 :)
세심한 리뷰 감사합니다 👍

eastjun-dev commented on js/Components/App.js:
```diff
@@ -0,0 +1,203 @@
+    });
+  };
+
+  // UserList
```
꼭 필요한 주석이 아니라면 삭제해주셔도 괜찮을 것 같아요 😄

2SOOY replied:
그렇군요! 해당 주석이 큰 의미를 갖고 있진 않네요 😢 삭제하겠습니다 :) 

@eastjun 님 하나 궁금한 점이 있는데, 동준님은 주석을 언제 활용하시나요 ?

eastjun-dev commented on js/Components/App.js:
```diff
@@ -0,0 +1,203 @@
+
+  // UserList
+  this.onChangeUser = async (userName) => {
+    const newState = await this.fetchState(userName);
```
비동기 통신의 경우 결과가 언제나 success일 수 없기 때문에 try~catch로 에러가 난 상황도 처리해주시면 더 안정적인 코드가 될 것 같아요 😄

2SOOY replied:
넵! `비동기에 대한 try~catch` 명심하겠습니다 :)

eastjun-dev commented on js/Components/TodoClearButton.js:
```diff
@@ -0,0 +1,34 @@
+  };
+
+  this.onClick = (e) => {
+    if (e.target.nodeName !== NODE.BUTTON) return;
```
nodeName으로 비교하시면, 아무래도 마크업이 변경되거나, 다른 버튼 ui가 생긴다면 원치 않은 에러가 발생할 수도 있을 것 같아요. 정확한 element를 찾기 위해 class를 추가해서 비교해보면 어떨까요?

2SOOY replied:
이후 마크업이 변경될 부분까지는 생각지도 못했네요 👍 
조금 더 `정확한 element 선택을 위해` 관련된 class까지 추가해서 조건을 타이트하게 검사할 수 있도록 하겠습니다 !

eastjun-dev commented on js/Components/TodoInput.js:
```diff
@@ -0,0 +1,45 @@
+  this.onKeyDown = (e) => {
+    if (e.key !== KEY.ENTER) return;
+
+    if (!this.name) {
```
꼼꼼한 예외처리가 돋보입니다! 👍

eastjun-dev commented on js/Components/TodoList.js:
```diff
@@ -0,0 +1,153 @@
+  };
+
+  this.onClick = (e) => {
+    const clickedClassName = e.target.className;
```
`className`으로 class를 가져와서 비교하는건 나중에 에러를 발생시킬 수 있을 것 같아요 😭 `className`의 경우 해당 element의 모든 class들을 string으로 return해주는데요. 만약 마크업상 스타일을 위해 다른 class가 추가된다면, 아래 if문들에서 원치 않는 동작을 일으킬 것 같아요 😭😭

2SOOY replied:
`className의 특성`에 대해서는 생각지도 못했던 부분이었네요 !
다른 class가 추가되었을 때, `className`으로 비교할 경우 반드시 문제가 생길 수 밖에 없는 로직이었군요 😢 

`classList.contains()`를 활용하여 수정하겠습니다 !

세심한 리뷰 감사합니다 👍👍

eastjun-dev commented on js/Components/TodoList.js:
```diff
@@ -0,0 +1,153 @@
+  };
+
+  this.createTodoListHTML = (todos) => {
+    return todos.reduce((html, todo) => {
```
`reduce`함수를 사용하셨군요 👍

2SOOY replied:
reduce에 익숙해지기 위해서 사용했습니다 :)

eastjun-dev commented on js/utils/api.js:
```diff
@@ -0,0 +1,95 @@
+    method: 'PUT',
+  }),
+
+  edit: (contents) => ({
```
`edit`과 `editPriority `의 경우 body에 담기는 데이터 외에 동일한 필드를 가지고 있으니 중복을 제거하고, 파라미터를 다르게해서 재사용하면 어떨까요? 🤔

2SOOY replied:
그러고보니 구조가 동일하군요 !
재사용 할 수 있도록 수정하겠습니다 👍

eastjun-dev commented on js/utils/api.js:
```diff
@@ -0,0 +1,95 @@
+};
+
+export const api = {
+  fetchUserList: () => {
```
array function의 경우 한 줄로 return하시는 경우는
`fetchUserList: () => request(`${BASE_URL}/`);`
위와 같이 간략하게 표현하는 것도 가능하답니다 😄

2SOOY replied:
이전부터 이 부분에 대해서는 리뷰를 듣곤 했었는데, 미처 적용하지 못했네요 
`() => ()` 적용하겠습니다 :)

eastjun-dev commented on js/utils/constant.js:
```diff
@@ -0,0 +1,63 @@
+  SELECTED_TAB: 'selectedTab',
+};
+
+export const MESSAGE = {
```
우와..... 너무 꼼꼼하세요 👍 👍 💯

2SOOY replied:
처음엔 번거로워도 나중되면 관리하는 게 편한 것 같아요 👍 
정말 감사합니다 😄

eastjun-dev commented on js/utils/constant.js:
```diff
@@ -0,0 +1,63 @@
+export const SELECTOR = {
```
여러개의 상수를 담고 있는 파일인 만큼 파일 명이 `constants.js`이면 어떨까요? 🤔

2SOOY replied:
복수형이 훨씬 적합한 네이밍이네요 ! 수정하겠습니다 :)

eastjun-dev reviewed (changes requested):
`validator`, `constants`등 수연님의 꼼꼼함이 돋보이는 PR이어서, 보는 내내 저도 너무 많은 학습을 할 수 있었어요 ㅎㅎ 좋은 코드 공유해주셔서 감사합니다! 더 안정성있는 코드를 위해 몇가지 피드백 남겨드렸는데 참고 부탁드릴게요 😉

2SOOY committed:
Refactor - TodoClearButton : check click event condition

- 리뷰 : https://github.com/next-step/js-todo-list-step2/pull/11#discussion_r476744320
- 단순히 nodeName으로 비교할 경우, 마크업 변경시 커버하지 못함
- 정확한 elem을 찾기 위한 class에 대한 조건을 추가함

2SOOY committed:
Style - TodoClearButton : add { } for return

- early reutrn을 위한 조건문이 길어질 경우
- 가독성을 위해서 { } 블록을 추가하여 return 사용

2SOOY committed:
Refactor - TodoList : clicke event check condition

- https://github.com/next-step/js-todo-list-step2/pull/11#discussion_r476748106
- 단순히 className을 비교할 경우
- 해당 elem가 다른 class를 갖고 있을 때, 문제가 생김
- 따라서 classList.contains()를 사용하여 class 포함 여부에 따른 로직 구성

2SOOY committed:
Style - constant.js to constants.js

- https://github.com/next-step/js-todo-list-step2/pull/11#discussion_r476758174
- constant.js 파일 내에서 여러가지 상수를 관리함
- 따라서 복수형의 네이밍이 더 적합함

2SOOY committed:
Refactor - TodoTab : change tab's classList logic

- https://github.com/next-step/js-todo-list-step2/pull/11#discussion_r476686303
- click 이벤트 발생시 e.target을 이용하여 자체적으로 classList 변경이 가능해짐
- 따라서 기존 render 메소드 자체가 필요없어짐
- 추가로 render 메소드 내 불필요한 조건도 없앨 수 있음

2SOOY committed:
Update - App : try...catch for api call

- 비동기 요청 시 try...catch 하는 부분이 빠져 있었음
- 안전한 동작을 위해 catch를 통한 error 콘솔로 보여주는 작업 추가
- 불필요한 주석 제거

read0more commented:
와...PR 내용도 그렇고 커밋 이력도 그렇고 README 내용도 그렇고 정리가 너무 잘돼있네요👍 코드도 깔끔하시고 코드를 보면서 덕분에 알게 된 부분도 있어서 잘봤습니다🙂 todo 수정할 때 커서를 어떻게 할까 고민을 했는데 select라는 메서드가 있었군요!
